### ISSUE TITLE
scalars don't raise overflow/underflow/divbyzero/invalid (migrated from Trac #1671)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#3222]
Original ticket http://projects.scipy.org/numpy/ticket/1671
Reported 2010-11-11 by atmention:mwiebe, assigned to unknown.

All these cases should be raising, I believe:

```
>>> import numpy as np
>>> err = np.seterr(all='raise')

>>> np.float32(1e-38) / np.float32(1e30)
0.0
>>> np.float32(1e-38) / np.array(1e38,np.float32)
0.0
>>> np.array(1e-38,np.float32) / np.float32(1e30)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
FloatingPointError: underflow encountered in divide

>>> np.float64(1e-300) / np.float64(1e300)
0.0
>>> np.float64(1e-300) / np.array(1e300, dtype=np.float64)
0.0
>>> np.array(1e-300, dtype=np.float64) / np.float64(1e300)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
FloatingPointError: underflow encountered in divide

>>> np.float32(1) / np.float32(0)
inf
>>> np.array(1,dtype=np.float32) / np.float32(0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
FloatingPointError: divide by zero encountered in divide

>>> np.inf-np.inf
nan
>>> np.array(np.inf)-np.inf
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
FloatingPointError: invalid value encountered in subtract
```

**Comment by thouis** [comment]
Comment in Trac by atmention:mwiebe, 2010-11-12

I dug in a little bit, and the problem appears to be the compiler optimizing too liberally.  If I add a printf between the operation and PyUFunc_getfperr() in scalarmathmodule.c.src, it works.

```
    atmention:nameatmention:_ctype_atmention:operatmention:(arg1, arg2, &out);
    printf("Result is %g\n", out);
#endif

#if atmention:fperratmention:
    /* Check status flag.  If it is set, then look up what to do */
    retstatus = PyUFunc_getfperr();
```

I'm guessing the compiler thinks it can reorder the float operation to after PyUFunc_getfperr, since PyUFunc_getfperr is independent of it except for the float exception state.

**Comment by thouis** [comment]
Comment in Trac by atmention:mwiebe, 2010-11-12

https://github.com/numpy/numpy/pull/13

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2011-03-30

Seems to be fixed now. The changes were merged in commit:72a702d22cfcdba03
